### ISSUE TITLE
Training score becomes NaN at epoch X

### ISSUE THREAD

**Issue by igor-93** [deeplearning4j/deeplearning4j#4001]
After Training for a while, I get Training score: NaN

#### Version Information
*nd4j-native-platform: 0.8.0
*deeplearning4j-core: 0.8.0
*dl4j-spark_2.11: 0.8.0_spark_2
* platform Information: Linux, 64 CPU-cores, used Spark and HDFS, no GPU

#### Code
`

	private void setHyperParams(){
		// scheduled lr decay: every 10 epochs, half the learning rate
		Map<Integer, Double> lrSchedule = new HashMap<>(); 
		lrSchedule.put(0, this.params.getLearningRate());
		lrSchedule.put(1, this.params.getLearningRate());
		lrSchedule.put(15*10, this.params.getLearningRate()/2.0);
		lrSchedule.put(15*20, this.params.getLearningRate()/4.0);
		lrSchedule.put(15*30, this.params.getLearningRate()/8.0);
		lrSchedule.put(15*40, this.params.getLearningRate()/16.0);
		
		double lrDecPower = 0.75;
		double lrDecRateInverse = 0.00004;
		
		confBuilder = new NeuralNetConfiguration.Builder()
				.seed(Parameters.rngSeed)
	            .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
	            .weightInit(WeightInit.XAVIER)
	            .iterations(Parameters.iterations)
	            .learningRate(this.params.getLearningRate())
	            .learningRateDecayPolicy(LearningRatePolicy.Schedule).learningRateSchedule(lrSchedule) 
	            //.learningRateDecayPolicy(LearningRatePolicy.Inverse).lrPolicyDecayRate(lrDecRateInverse).lrPolicyPower(lrDecPower)
	            
	            .updater(Updater.ADAM)
	            
	            .regularization(true)
	            .list();
	}
	
	/**
	 * Builds the network. Here is where we define the architecture of the CNN.
	 */
	private void setCNNArch(){
		log.info("Build the model...");
		cnn = confBuilder
				.layer(0, new ConvolutionLayer.Builder(5,5)
						.nIn(Parameters.numChannels)
						.stride(1,1)
						.nOut(64)
						.activation(Activation.ELU)
						.weightInit(WeightInit.XAVIER)
						.dropOut(Parameters.DEF_DROPOUT_RATIO)
						.build())
				
				.layer(1, new BatchNormalization.Builder().build())
				
				.layer(2, buildPoolingLayer())
				
				.layer(3, buildConvLayer(3, 128))
				
				.layer(4, new BatchNormalization.Builder().build())
				
				.layer(5, buildPoolingLayer())
				
				.layer(6, buildConvLayer(3, 128))
				
				.layer(7, new BatchNormalization.Builder().build())
				
				//.layer(8, buildPoolingLayer())
				
				//.layer(6, buildConvLayer(3, 512))
				
				//.layer(7, buildPoolingLayer())
				
				.layer(6, new DenseLayer.Builder()
						.nOut(300)							
						.activation(Activation.ELU)
						.weightInit(WeightInit.XAVIER)
						//.l2(this.params.getL2reg())
						.dropOut(Parameters.DEF_DROPOUT_RATIO)
						.build())
			
				.layer(7, new DenseLayer.Builder()
						.nOut(150)							
						.activation(Activation.ELU)
						.weightInit(WeightInit.XAVIER)
						//.l2(this.params.getL2reg())
						.dropOut(Parameters.DEF_DROPOUT_RATIO)
						.build())
				
				.layer(8, new OutputLayer.Builder(LossFunction.XENT)
						.nOut(Parameters.nClasses)
						.activation(Activation.SOFTMAX)
						.weightInit(WeightInit.XAVIER)
						.build())
				.setInputType(InputType.convolutional(inputImgSize, inputImgSize, Parameters.numChannels))
				
				.backprop(true).pretrain(false).build();
	}
	
	/**
	 * Builds standard Max-pooling layer.
	 * @return the defined layer
	 */
	private SubsamplingLayer buildPoolingLayer(){
		return new SubsamplingLayer.Builder(SubsamplingLayer.PoolingType.MAX)
				.kernelSize(Parameters.poolingFactor, Parameters.poolingFactor)
				.stride(Parameters.poolingFactor, Parameters.poolingFactor)
				.convolutionMode(ConvolutionMode.Same)
				.build();
	}
	
	/**
	 * Builds standard convolutional layer for the given kernel size and with (nOut).
	 * @return the defined layer
	 */
	private ConvolutionLayer buildConvLayer(int kernelSize, int nOut){
		return new ConvolutionLayer.Builder(kernelSize, kernelSize)
				.stride(1,1)
				.nOut(nOut)
				.weightInit(WeightInit.XAVIER)
				.activation(Activation.ELU)
				.convolutionMode(ConvolutionMode.Same)
				.dropOut(Parameters.DEF_DROPOUT_RATIO)
				.build();
	}
	
	
	/**
	 * Main function of the class. runs either training of prediction.
	 * @throws IOException
	 */
	public void run() throws IOException{
		if(mode == Mode.TRAIN)
			runTrain();
		if(mode == Mode.PREDICT)
			runPredict();
	}
	
	/**
	 * Predicts the classes for the provided images.
	 */
	private void runPredict() {
		String path = this.modelDir + File.separator + this.modelFileName;
		MultiLayerNetwork restored = ioops.loadNetwork(path);
		
		NativeImageLoader loader = new NativeImageLoader(inputImgSize, inputImgSize, Parameters.numChannels);
		
		List<String> imgFiles = ioops.filterImgs(predictDataDir);
		
		for(String imgFile : imgFiles) {
			INDArray image = ioops.loadImg(imgFile, loader);
			INDArray output = restored.output(image);
			log.info("List of probabilities per label:");
			log.info(output.toString());
		}
	}
	
	/**
	 * Trains the model on Spark. Also saves the model to the file and if
	 * evaluation is enabled it evaluates the model in each epoch on the test data and saves the ntwork statistics.
	 * These can later be visualized with detector-stats-ui.
	 * @throws IOException
	 */
	private void runTrain() throws IOException{
		loadData();
		String exportDirectory = "directory-temp";
		
		@SuppressWarnings("rawtypes")
		ParameterAveragingTrainingMaster tm = new ParameterAveragingTrainingMaster.Builder(1)
				.averagingFrequency(1)		// frequency in which params are averaged
				.workerPrefetchNumBatches(2)
				.batchSizePerWorker(this.params.getBatchSizePerWorker())
				.exportDirectory(exportDirectory)
				.saveUpdater(true)
				
				.build();
		tm.setCollectTrainingStats(true);
	
		
		SparkDl4jMultiLayer sparkNet = new SparkDl4jMultiLayer(sc, cnn, tm);
		sparkNet.setCollectTrainingStats(true);
		
		File statsFile = new File(statsFileName);
		StatsStorage ss = new FileStatsStorage(statsFile);
		log.info("Storing net stats in "+statsFile.getAbsolutePath());
		
		List<IterationListener> listeners = new ArrayList<IterationListener>();
		StatsListener sl = new StatsListener(null);
		TrainingListener tl = new MyTrainingListener();
		String convSaveFolder = ioops.getModelFolder() + File.separator + "convs"; 
		//OfflineConvolutionalIterationListener convListener = new OfflineConvolutionalIterationListener(convSaveFolder, 1000);
		listeners.add(sl);
		//listeners.add(tl);
		//listeners.add(convListener);
		if(runEval){
			sparkNet.setListeners(ss, listeners);
		}
	    
		log.info("Train the model...");
		for(int i = 0; i < this.params.getNumEpochs(); i++){
			log.info("Running epoch " + i);
			sparkNet.fit(trainData);
			
			log.info("Iteration count after epoch: "+cnn.getIterationCount());
			log.info("Training score: "+sparkNet.getScore());
			
			if(runEval && i % 2 == 0){
				Evaluation eval_train = sparkNet.evaluate(trainData);
				Evaluation eval_test = sparkNet.evaluate(testData);
				log.info("***** Evaluation Train *****");
				log.info(eval_train.stats());
				log.info("***** Evaluation Test *****");
				log.info(eval_test.stats());
			}
			
			if(i % this.params.getSavingFreq() == 0){
				String locationToSave = this.modelDir + File.separator + this.modelFileName;
				boolean saveUpdater = true; 
				ioops.saveNetwork(locationToSave, sparkNet.getNetwork(), saveUpdater);
			}
		}
		
		log.info("Training finished!");		
		
		tm.deleteTempFiles(sc);
	}
	
	/**
	 * Loads training (and test) data. Data is loaded from HDFS to Spark RDDs.
	 * The data is also scaled, randomly split into test/train and labels are generated from the parent folder name.  
	 * @throws IOException
	 * @throws IllegalStateException
	 */
	private void loadData() throws IOException, IllegalStateException {
		BaseInputSplit filesInDir = new CustomFileSplit(ioops, Parameters.randNumGen);
		PathLabelGenerator labelMaker;
		if(local){
			labelMaker = new ParentPathLabelGenerator();
		}
		else {
			labelMaker = new HDFSParentPathLabelGenerator();
		}
		
		BalancedPathFilter pathFilter = new BalancedPathFilter(Parameters.randNumGen, allowedExtensions, labelMaker);
		
		if(!runEval){
			this.params.setTestSize(0);
		}
		
		InputSplit[] filesDirSplit = filesInDir.sample(pathFilter, 100 - this.params.getTestSize(), this.params.getTestSize());
		InputSplit trainDataSplit = filesDirSplit[0];
		InputSplit testDataSplit = filesDirSplit[1];
		log.info("All data paths: "+ Long.valueOf(filesInDir.locations().length));
		log.info("Train data paths: "+ Long.valueOf(trainDataSplit.locations().length));
		log.info("Test data paths: "+Long.valueOf(testDataSplit.locations().length));
		
		// Read files
		JavaPairRDD<String, PortableDataStream> origData = sc.binaryFiles(inputDataDir + File.separator + "*");
		log.info("All data: "+ Long.valueOf(origData.count()));
		
		log.info("trainDataSplit.locations()");
		for(int i = 0; i < trainDataSplit.locations().length; i++){
			log.info("  "+trainDataSplit.locations()[i]);
		}
		
		log.info("testDataSplit.locations()");
		for(int i = 0; i < testDataSplit.locations().length; i++){
			log.info("  "+testDataSplit.locations()[i]);
		}
		// filter origData so it contains only trainDataSplit.locations()
		JavaPairRDD<String, PortableDataStream> origDataTrain = origData.filter(new FilterSplitsFunction(trainDataSplit));
			
		//Need this for Spark: can't infer without init call
		ImageRecordReader recordReaderTrain = new ImageRecordReader(inputImgSize, inputImgSize, Parameters.numChannels, labelMaker);
		recordReaderTrain.setLabels(labelsList);
		JavaRDD<List<Writable>> rddTr = origDataTrain.map(new RecordReaderFunction(recordReaderTrain));
		this.trainData = rddTr.map(new DataVecDataSetFunction(1, labelsList.size(), false));
		this.trainData = this.trainData.map(new DataSetScalerFunction(scaler));
		log.info("Train data: "+ Long.valueOf(trainData.count()));
		
		if(recordReaderTrain.numLabels() != Parameters.nClasses){
			int labelsFound = recordReaderTrain.numLabels();
			recordReaderTrain.close();
			throw new IllegalStateException("Num labels doesnt match! "+labelsFound +" vs "+Parameters.nClasses);
		}
		
		if(runEval){
			// filter origData so it contains only testDataSplit.locations()
			JavaPairRDD<String, PortableDataStream> origDataTest = origData.filter(new FilterSplitsFunction(testDataSplit));
			
			ImageRecordReader recordReaderTest = new ImageRecordReader(inputImgSize, inputImgSize, Parameters.numChannels, labelMaker);
			recordReaderTest.setLabels(labelsList);
			JavaRDD<List<Writable>> rddTe = origDataTest.map(new RecordReaderFunction(recordReaderTest));
			this.testData = rddTe.map(new DataVecDataSetFunction(1, 2, false));
			this.testData = this.testData.map(new DataSetScalerFunction(scaler));
			log.info("Test data: "+ Long.valueOf(testData.count()));
		}
	}`

###Output:
17/09/04 14:32:53 INFO DeepClassifier: Running epoch 15
17/09/04 14:36:58 INFO DeepClassifier: Iteration count after epoch: 208
17/09/04 14:36:58 INFO DeepClassifier: Training score: 0.40788027085363865
17/09/04 14:36:58 INFO DeepClassifier: Running epoch 16
17/09/04 14:40:57 INFO DeepClassifier: Iteration count after epoch: 221
17/09/04 14:40:57 INFO DeepClassifier: Training score: 0.5098298899829388
17/09/04 14:40:57 WARN TaskSetManager: Stage 700 contains a task of very large size (515 KB). The maximum recommended task size is 100 KB.
^[[A17/09/04 14:43:57 WARN TaskSetManager: Stage 701 contains a task of very large size (515 KB). The maximum recommended task size is 100 KB.
17/09/04 14:45:18 INFO DeepClassifier: ***** Evaluation Train *****
17/09/04 14:45:18 INFO DeepClassifier:
Examples labeled as 0 classified by model as 0: 1487 times
Examples labeled as 0 classified by model as 1: 249 times
Examples labeled as 1 classified by model as 0: 250 times
Examples labeled as 1 classified by model as 1: 1488 times

==========================Scores=====================================
 Accuracy:        0.8564
 Precision:       0.8564
 Recall:          0.8564
 F1 Score:        0.8564
===================================================================
17/09/04 14:45:18 INFO DeepClassifier: ***** Evaluation Test *****
17/09/04 14:45:18 INFO DeepClassifier:
Examples labeled as 0 classified by model as 0: 150 times
Examples labeled as 0 classified by model as 1: 44 times
Examples labeled as 1 classified by model as 0: 48 times
Examples labeled as 1 classified by model as 1: 144 times

==========================Scores========================================
 Accuracy:        0.7617
 Precision:       0.7618
 Recall:          0.7616
 F1 Score:        0.7617
===================================================================
17/09/04 14:45:18 INFO DeepClassifier: Running epoch 17
17/09/04 14:49:18 INFO DeepClassifier: Iteration count after epoch: 234
17/09/04 14:49:18 INFO DeepClassifier: Training score: 0.5366226402111351
17/09/04 14:49:18 INFO DeepClassifier: Running epoch 18
17/09/04 14:53:13 INFO DeepClassifier: Iteration count after epoch: 247
17/09/04 14:53:13 INFO DeepClassifier: Training score: NaN
17/09/04 14:53:13 WARN TaskSetManager: Stage 782 contains a task of very large size (515 KB). The maximum recommended task size is 100 KB.
17/09/04 14:56:18 WARN TaskSetManager: Stage 783 contains a task of very large size (515 KB). The maximum recommended task size is 100 KB.
17/09/04 14:57:49 INFO DeepClassifier: ***** Evaluation Train *****
17/09/04 14:57:49 INFO DeepClassifier:
Examples labeled as 0 classified by model as 0: 1736 times
Examples labeled as 1 classified by model as 0: 1738 times

Warning: class 1 was never predicted by the model. This class was excluded from the average precision

==========================Scores=====================================
 Accuracy:        0.4997
 Precision:       0.4997
 Recall:          0.5
 F1 Score:        0.4999
===================================================================
17/09/04 14:57:49 INFO DeepClassifier: ***** Evaluation Test *****
17/09/04 14:57:49 INFO DeepClassifier:
Examples labeled as 0 classified by model as 0: 194 times
Examples labeled as 1 classified by model as 0: 192 times

Warning: class 1 was never predicted by the model. This class was excluded from the average precision

==========================Scores=====================================
 Accuracy:        0.5026
 Precision:       0.5026
 Recall:          0.5
 F1 Score:        0.5013
===================================================================

**Comment by AlexDBlack** [comment]
First thing: I'd recommend upgrading to 0.9.1

Second: usually NaN is due to tuning. I'd recommend reading these two pages:
https://deeplearning4j.org/troubleshootingneuralnets
https://deeplearning4j.org/visualization

Third: you are using Spark on a single machine? Don't. It's much less efficient than using ParallelWrapper for this purpose.

If you are still having problems after that, we can reopen this.
